design/20220118.server-side-apply.md

```diff
@@ -0,0 +1,156 @@
+operations result in resource version conflicts. This results in error logs
+that regularly confuse users and miss-attribute the problem they are having, as
+well as needless re-queuing churn through the controller slowing cert-manager
+down.
+
+### Deleting Fields
+
+Without Server Side Apply, it is difficult for cert-manager to reason about some
+fields that should be _deleted_ on an UPDATE operation during reconciliation.
+One such example are Certificate SecretTemplate's Annotations and Labels. If a
+key was removed from either field on the Certificate, the next reconciliation
+would not know which keys should be preserved or removed (as they may have been
+annotated or labelled from a third party).
+
+Managed fields define ownership of these fields, therefore cert-manager is able
+to observe a discrepancy due to a previous Certificate state and perform another
+apply (omitting those keys), resulting in those fields being removed. Without
+managed fields, cert-manager would have to create a third state store in either
+annotations or status fields on that resource.
+
+
+## Field Manager and User Agent
+
+Field Manager is a string denoting the manager who owns a field or set of fields
+on a resource. This string is set by the client at API call time and is
+derived by either the client setting the `fieldManager` URL query explicitly, or
+calculated from the client's user agent ([characters preceding the first /, quote
+unprintable character and then trim what's beyond the 128
+limit](https://github.com/kubernetes/kubernetes/blob/9a75e7b0fd1b567f774a3373be640e19b33e7ef1/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L252)).
+
+cert-manager will use a consistent naming scheme for both the user agent prefix
+and field manager across all components. Each component will have a human
+readable name that describes that component, and is used for both its user agent
+and field manager. Using the same name enables better auditing trails that helps
+debugging and improves telemetry.
+
+### Scheme
+
+### Field Manager
+
+The field manager is derived from the component in cert-manager, prefixed by the
+string `cert-manager`:
+
+```text
+cert-manager-<cert-manager component>
+```
+
+e.g.
+
+```text
+cert-manager-certificates-issuing
+```
+
+### User Agent
+
+The user agent is inspired by the [client-go default user
+agent](https://github.com/kubernetes/client-go/blob/664b1a6c8ce9c92ce65bef3f9833b402449c98d2/rest/config.go#L499),
+and is defined with the following:
+
+```text
+<component field manager>/<cert-manager version> (<Operating System>/<Architecture>) cert-manager/<git commit>
+```
+
+e.g:
+
+```text
+cert-manager-certificates-issuing/v1.7.0 (Linux/amd64) cert-manager/0b686b8f38c8c7442744c9224d18e780ee7f244a
+```
+
+## Implementation Consideration
+
+All CREATE and UPDATE operations in cert-manager controllers are to be replaced
+with APPLY. For this change, controllers will need to be modified so that for
```

Comment:
Question: will this change how our controllers reconcile on status updates? (This may have been discussed before, but assuming that some control loop is currently getting triggered by update events to a `Certificate`'s status, will this still function the same way with APPLYs?)

Reply:
Yes, the informers will all behave the same way with APPLY.